src/lib.rs

```diff
@@ -462,41 +481,120 @@ where
+            if symbol.file.is_some() || self.opts.with_functions {
+                let file = symbol
+                    .file
+                    .map(|file| path::PathBuf::from(&*String::from_utf8_lossy(file)));
+                let name = demangle_any_symbol(symbol.name)
+                    .or_else(|| Some(String::from_utf8_lossy(symbol.name)));
+                return Ok(Some((file, None, name)));
+            }
+        }
+
         Ok(None)
     }
 }
 
+/// Demangle a symbol when we don't know which language it is.
+fn demangle_any_symbol(mangled: &[u8]) -> Option<Cow<str>> {
+    demangle_cpp_symbol(mangled).or_else(|| demangle_rust_symbol(mangled))
+}
+
 #[cfg(feature = "cpp_demangle")]
-fn demangle_cpp_symbol(mangled: &[u8]) -> Cow<str> {
-    if let Ok(sym) = cpp_demangle::Symbol::new(mangled) {
-        Cow::from(format!("{}", sym))
-    } else {
-        String::from_utf8_lossy(mangled)
-    }
+fn demangle_cpp_symbol(mangled: &[u8]) -> Option<Cow<str>> {
+    cpp_demangle::Symbol::new(mangled)
+        .ok()
+        .map(|sym| Cow::from(format!("{}", sym)))
 }
 
 #[cfg(not(feature = "cpp_demangle"))]
-fn demangle_cpp_symbol(mangled: &[u8]) -> Cow<str> {
-    String::from_utf8_lossy(mangled)
+fn demangle_cpp_symbol(mangled: &[u8]) -> Option<Cow<str>> {
+    None
 }
 
 #[cfg(feature = "rustc-demangle")]
-fn demangle_rust_symbol(mangled: &[u8]) -> Cow<str> {
-    Cow::from(format!(
+fn demangle_rust_symbol(mangled: &[u8]) -> Option<Cow<str>> {
+    Some(Cow::from(format!(
         "{}",
         rustc_demangle::demangle(String::from_utf8_lossy(mangled).as_ref())
-    ))
+    )))
 }
 
 #[cfg(not(feature = "rustc-demangle"))]
-fn demangle_rust_symbol(mangled: &[u8]) -> Cow<str> {
-    String::from_utf8_lossy(mangled)
+fn demangle_rust_symbol(mangled: &[u8]) -> Option<Cow<str>> {
+    None
+}
+
+struct Symbol<'input> {
+    name: &'input [u8],
+    file: Option<&'input [u8]>,
+    begin: u64,
+    end: u64,
+}
+
+fn parse_symbols<'input>(file: &object::File<'input>) -> Vec<Symbol<'input>> {
+    let mut symbols = Vec::new();
+    let mut filename = None;
+    for symbol in file.get_symbols() {
+        match symbol.kind() {
+            SymbolKind::Unknown | SymbolKind::Text | SymbolKind::Data => {}
+            SymbolKind::File => {
+                if symbol.name().len() == 0 {
+                    filename = None;
+                } else {
+                    filename = Some(symbol.name());
+                }
+                continue;
+            }
+            SymbolKind::Section | SymbolKind::Common | SymbolKind::Tls => continue,
+        }
+        if symbol.is_undefined() || symbol.name().len() == 0 || symbol.size() == 0 {
+            continue;
+        }
+        symbols.push(Symbol {
+            name: symbol.name(),
+            file: filename,
+            begin: symbol.address(),
+            end: symbol.address() + symbol.size(),
+        });
+    }
+
+    symbols.sort_by(|a, b| {
```

Comment:
Would it perhaps be an idea to insert the symbols into a `BTreeMap` instead? I believe that should be faster.

Reply:
I don't think we want to use `BTreeMap`. I think what we want is a different array layout (see https://arxiv.org/abs/1509.05053). I'm not aware of an implementation of this in rust though.

Reply:
I guess the question here is what we're expecting `n` to be, and the expected number of searches. Figure 17 of that paper is likely the comparison we want. When `n` is large (as in > =2^18, or perhaps smaller depending on the efficiency of Rust's binary search implementation), or when the number of lookups is small, I believe a `BTreeMap` is the appropriate structure until someone does implement Eytzinger (the former is supported by the paper, the latter by the fact that the array also needs to be sorted in order to do binary searches). I suspect that we expect `n` to be fairly small, though I don't have a good sense of the number of symbols in a representative symbol file. It'd be useful if the paper above also included the cost of sorting, but a cursory reading suggests that they don't.

Without an implementation of Eytzinger (which seems like it might be interesting), I think `BTreeMap` is still the right way to go. For small `n`, latency will be pretty short anyway, and for large `n`, `BTreeMap` wins.

Reply:
Reading a bit more closely, their B-tree isn't anywhere near the same as a Rust `BTreeMap`. It assume contiguous storage. Hard to say exactly how the two would compare, but Rust's would likely be slower. I guess then I'm fine with the current code, even though I feel like the sorting will actually introduce a non-trivial up-front cost that might not be right trade-off if we expect few lookups.

Reply:
@philipc fyi: https://github.com/patmorin/arraylayout/issues/3

Reply:
Yes, I meant we want to use contiguous storage as currently, but with a better layout and search algorithm than binary search, so not rust's `BTreeMap` because it is not contiguous. Whether that sorting strategy should be B-tree or Eytzinger needs testing.

> I guess then I'm fine with the current code, even though I feel like the sorting will actually introduce a non-trivial up-front cost that might not be right trade-off if we expect few lookups.

So I'm not familiar with the costs of `BTreeMap` construction, but I didn't expect it to be significantly better than a sort?

Reply:
I should also point out that the `object` crate is already sorting Mach-O symbols internally to calculate symbol sizes (but it doesn't give us the sorted array), and there's a bunch of inefficiencies due to creating multiple arrays of these symbols (eg for ELF there's one in the `goblin` crate, one in the `object` crate and one here) so that needs optimizing too. Ideally `goblin` and `object` would both give iterators instead, but I don't know how to get that to work with the Mach-O symbol sizes.

Reply:
> So I'm not familiar with the costs of `BTreeMap` construction, but I didn't expect it to be significantly better than a sort?
I believe it should be faster, as you avoid walking all the elements once at the beginning to push them onto the list. The sorting itself is probably roughly the same, although that depends on the exact algorithm Rust uses internally (since it might shift things in the array). You might also want to use `sort_unstable_by` here.

Ah, I wasn't aware of the sorting in `object`. I don't know which of these approaches are best suited for adopting an already-sorted array, though it might very well be `push` + `sort_by`.

Reply:
@philipc First implementation here: https://github.com/jonhoo/ordsearch

Reply:
Nice.  I'm going to leave optimisation of this for future work.